query-algebrizer/src/cc.rs

```diff
@@ -917,4 +975,75 @@ mod testing {
         ]);
     }
 
+    #[test]
+    fn test_known_bindings() {
+        let mut schema = Schema::default();
+
+        associate_ident(&mut schema, NamespacedKeyword::new("foo", "bar"), 99);
+        add_attribute(&mut schema, 99, Attribute {
+            value_type: ValueType::Boolean,
+            ..Default::default()
+        });
+
+        let x = Variable(PlainSymbol::new("?x"));
+        let y = Variable(PlainSymbol::new("?y"));
+
+        let b: BTreeMap<Variable, TypedValue> =
+            vec![(y.clone(), TypedValue::Boolean(true))].into_iter().collect();
+        let mut cc = ConjoiningClauses::with_bindings(b);
+
+        cc.apply_pattern(&schema, &Pattern {
+            source: None,
+            entity: PatternNonValuePlace::Variable(x.clone()),
+            attribute: PatternNonValuePlace::Ident(NamespacedKeyword::new("foo", "bar")),
+            value: PatternValuePlace::Variable(y.clone()),
+            tx: PatternNonValuePlace::Placeholder,
+        });
+
+        let d0_e = QualifiedAlias("datoms00".to_string(), DatomsColumn::Entity);
+        let d0_a = QualifiedAlias("datoms00".to_string(), DatomsColumn::Attribute);
+        let d0_v = QualifiedAlias("datoms00".to_string(), DatomsColumn::Value);
+
+        // ?y has been expanded into `true`.
+        assert_eq!(cc.wheres, vec![
+                   ColumnConstraint::EqualsEntity(d0_a, 99),
+                   ColumnConstraint::EqualsValue(d0_v, TypedValue::Boolean(true)),
+        ]);
+
+        // There is no binding for ?y.
+        assert!(!cc.column_bindings.contains_key(&y));
+
+        // ?x is bound to the entity.
+        assert_eq!(cc.column_bindings.get(&x).unwrap(),
+                   &vec![d0_e.clone()]);
+    }
+
+    #[test]
+    fn test_known_bindings_type_disagreement() {
+        let mut schema = Schema::default();
+
+        associate_ident(&mut schema, NamespacedKeyword::new("foo", "bar"), 99);
+        add_attribute(&mut schema, 99, Attribute {
+            value_type: ValueType::Boolean,
+            ..Default::default()
+        });
+
+        let x = Variable(PlainSymbol::new("?x"));
+        let y = Variable(PlainSymbol::new("?y"));
+
+        let b: BTreeMap<Variable, TypedValue> =
+            vec![(y.clone(), TypedValue::Long(42))].into_iter().collect();
+        let mut cc = ConjoiningClauses::with_bindings(b);
+
+        cc.apply_pattern(&schema, &Pattern {
+            source: None,
+            entity: PatternNonValuePlace::Variable(x.clone()),
+            attribute: PatternNonValuePlace::Ident(NamespacedKeyword::new("foo", "bar")),
+            value: PatternValuePlace::Variable(y.clone()),
+            tx: PatternNonValuePlace::Placeholder,
+        });
+
+        // The type of the provided binding doesn't match the type of the attribute.
```

Comment:
You had a mechanism, IIRC, for explaining _why_ the clauses are empty.  Can you assert something about this?  It's valuable to provide such type-checking to consumers.